application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/TrialLicenseGenerator.java

```diff
@@ -0,0 +1,251 @@
+import org.xwiki.properties.converter.Converter;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.xpn.xwiki.XWikiContext;
+import com.xwiki.licensing.License;
+import com.xwiki.licensing.LicenseManager;
+import com.xwiki.licensing.LicensedExtensionManager;
+import com.xwiki.licensing.LicensingConfiguration;
+import com.xwiki.licensing.Licensor;
+
+/**
+ * Helper methods for generating a trial license and updating it.
+ *
+ * @since 1.17
+ * @version $Id$
+ */
+@Component(roles = TrialLicenseGenerator.class)
+@Singleton
+public class TrialLicenseGenerator
+{
+    private static final String FAILED_TO_ADD_TRIAL_LICENSE = "Failed to add trial license";
+
+    private static final String FEATURE_ID = "featureId";
+
+    private static final String INSTANCE_ID = "instanceId";
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private Provider<InstanceIdManager> instanceIdManagerProvider;
+
+    @Inject
+    private UserCounter userCounter;
+
+    @Inject
+    private LicensedExtensionManager licensedExtensionManager;
+
+    @Inject
+    private LicensingConfiguration licensingConfig;
+
+    @Inject
+    private Provider<Licensor> licensorProvider;
+
+    @Inject
+    private Provider<LicenseManager> licenseManagerProvider;
+
+    @Inject
+    @Named("Base64")
+    private BinaryStringEncoder base64decoder;
+
+    @Inject
+    private Converter<License> converter;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    /**
+     * Generate trial license for the given extension.
+     *
+     * @param extensionId the extension for which to generate a trial license
+     */
+    public void generateTrialLicense(ExtensionId extensionId)
+    {
+        try {
+            URL trialURL = getTrialURL(extensionId);
+            if (trialURL == null) {
+                logger.info(FAILED_TO_ADD_TRIAL_LICENSE);
+                return;
+            }
+
+            XWikiContext xcontext = contextProvider.get();
+            String getTrialResponse = xcontext.getWiki().getURLContent(trialURL.toString(), xcontext);
+
+            if (getTrialResponse.contains("error")) {
+                logger.info(FAILED_TO_ADD_TRIAL_LICENSE);
+            } else {
+                logger.info("Added trial license");
+                updateLicenses();
+            }
+        } catch (Exception e) {
+            logger.warn("Failed to get trial license for [{}]. Root cause is [{}]", extensionId,
+                ExceptionUtils.getRootCauseMessage(e));
+        }
+    }
+
+    /**
+     * Check if the given extension is a mandatory licensed extension and the licensing owner information is complete.
+     *
+     * @param extensionId extension to be checked
+     * @return true if a trial license can be generated for the given extension, false otherwise
+     */
+    public Boolean canGenerateTrialLicense(ExtensionId extensionId)
+    {
+        return isOwnerDataComplete() && isMandatoryLicensedExtension(extensionId);
```

Comment:
Shouldn't we also check if there is already an active license for the specified extension? Otherwise we will try to generate a trial license when upgrading a paid app for which there is a license already. No?